File: skills/create-cuebook-content/references/skills/render-cuebook-viewpoint-visual/scripts/finalize_wide_viewpoint.py
#!/usr/bin/env python3
"""Freeze a selected VisualDirectionSetV1 direction as a wide ViewpointVisualV1."""

from __future__ import annotations

import argparse
import hashlib
import importlib.util
import json
import os
import re
import sys
from pathlib import Path, PurePosixPath
from typing import Any


ROOT = Path(__file__).resolve().parents[1]
SKILLS_ROOT = ROOT.parent
sys.path.insert(0, str(ROOT / "scripts"))
from validate_viewpoint_visual import validate_manifest  # noqa: E402


def load_direction_validator():
    path = SKILLS_ROOT / "direct-cuebook-viewpoint-visual" / "scripts" / "validate_visual_direction_set.py"
    spec = importlib.util.spec_from_file_location("cuebook_direction_validator", path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Cannot load direction validator: {path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def sha256_bytes(data: bytes) -> str:
    return "sha256:" + hashlib.sha256(data).hexdigest()


def safe_asset_path(asset_root: Path, ref: str, *, base: Path | None = None) -> Path:
    relative = Path(ref)
    if relative.is_absolute() or ".." in relative.parts:
        raise RuntimeError(f"Unsafe artifact-local asset ref: {ref}")
    root = asset_root.resolve()
    candidate = ((base or asset_root) / relative).resolve()
    try:
        candidate.relative_to(root)
    except ValueError as exc:
        raise RuntimeError(f"Asset ref escapes the artifact root: {ref}") from exc
    return candidate


def production_font_manifest(html_text: str, asset_root: Path) -> tuple[str, bytes]:
    def attribute(name: str) -> str:
        match = re.search(rf'\b{re.escape(name)}=["\']([^"\']+)["\']', html_text, flags=re.I)
        if not match:
            raise RuntimeError(f"Selected HTML is missing {name}.")
        return match.group(1)

    if attribute("data-font-profile") != "cuebook-noi-v1":
        raise RuntimeError("Selected HTML must use the cuebook-noi-v1 font profile.")
    if attribute("data-font-license-mode") != "production":
        raise RuntimeError("Final publication requires production font license mode.")
    manifest_ref = attribute("data-font-manifest-ref")
    manifest_path = safe_asset_path(asset_root, manifest_ref)
    if not manifest_path.is_file():
        raise RuntimeError(f"Font manifest is missing: {manifest_path}")
    manifest_bytes = manifest_path.read_bytes()
    try:
        manifest = json.loads(manifest_bytes)
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"Font manifest is invalid JSON: {manifest_path}") from exc
    if manifest.get("schema_version") != "cuebook-font-assets-v1" or manifest.get("font_profile_id") != "cuebook-noi-v1":
        raise RuntimeError("Font manifest does not bind the Cuebook Noi profile.")
    license_ref = str(manifest.get("license_ref") or "")
    if manifest.get("license_mode") != "production" or manifest.get("release_eligible") is not True:
        raise RuntimeError("Font manifest is not release-eligible production material.")
    if len(license_ref) < 6 or re.search(r"trial|eval", license_ref, flags=re.I):
        raise RuntimeError("Production font manifest needs an opaque non-evaluation license_ref.")
    css_ref = str(manifest.get("css_ref") or "")
    css_path = safe_asset_path(asset_root, css_ref, base=manifest_path.parent)
    if not css_path.is_file() or sha256_bytes(css_path.read_bytes()) != manifest.get("css_sha256"):
        raise RuntimeError("Font CSS is missing or does not match the manifest hash.")
    expected_css_ref = (PurePosixPath(manifest_ref).parent / css_ref).as_posix()
    if not re.search(rf'href=["\'](?:\./)?{re.escape(expected_css_ref)}["\']', html_text, flags=re.I):
        raise RuntimeError("Selected HTML does not load the CSS bound by its font manifest.")
    files = manifest.get("files")
    if not isinstance(files, list) or {item.get("weight") for item in files if isinstance(item, dict)} != {400, 500, 600, 700}:
        raise RuntimeError("Font manifest must bind upright Noi weights 400, 500, 600, and 700.")
    for item in files:
        ref = str(item.get("ref") or "")
        source_name = str(item.get("source_name") or "")
        if re.search(r"trial", ref + source_name, flags=re.I):
            raise RuntimeError("Production font manifest cannot reference Trial font assets.")
        font_path = safe_asset_path(asset_root, ref, base=manifest_path.parent)
        if not font_path.is_file() or sha256_bytes(font_path.read_bytes()) != item.get("sha256"):
            raise RuntimeError(f"Font asset is missing or does not match its manifest hash: {ref}")
    return manifest_ref, manifest_bytes


def unique(values: list[str]) -> list[str]:
    return list(dict.fromkeys(value for value in values if value))


def selected_direction(direction_set: dict[str, Any]) -> dict[str, Any]:
    if direction_set.get("state") != "selected":
        raise RuntimeError("VisualDirectionSetV1 must be selected before finalization.")
    selected_id = direction_set.get("selected_direction_id")
    for direction in direction_set.get("directions", []):
        if direction.get("direction_id") == selected_id:
            return direction
    raise RuntimeError("Selected direction does not resolve inside VisualDirectionSetV1.")


def payload_mode(bindings: list[dict[str, Any]]) -> str:
    kinds = {item.get("kind") for item in bindings}
    has_series = "series" in kinds
    has_key_numbers = bool(kinds.intersection({"metric", "level"})) or any(
        isinstance(item.get("value"), (int, float)) and not isinstance(item.get("value"), bool)
        for item in bindings
    )
    has_qualitative = bool(kinds - {"series", "metric", "level"})
    if has_series and (has_key_numbers or has_qualitative):
        return "mixed"
    if has_series:
        return "series"
    if has_key_numbers:
        return "key_numbers"
    return "qualitative"


def build_manifest(
    direction_set: dict[str, Any],
    asset_root: Path,
    *,
    observed_at: str,
    decision_cutoff_at: str,
    generated_at: str,
    state: str = "frozen",
) -> dict[str, Any]:
    direction = selected_direction(direction_set)
    binding_ids = set(direction.get("binding_refs", []))
    bindings = [item for item in direction_set.get("bindings", []) if item.get("binding_id") in binding_ids]
    if not bindings:
        raise RuntimeError("Selected direction has no resolved bindings.")

    html_path = asset_root / direction["html_ref"]
    full_path = asset_root / direction["preview_ref"]
    compact_path = asset_root / direction["compact_preview_ref"]
    capture_path = asset_root / direction["capture_report_ref"]
    for path in (html_path, full_path, compact_path, capture_path):
        if not path.is_file():
            raise RuntimeError(f"Required selected-direction asset is missing: {path}")

    html_bytes = html_path.read_bytes()
    font_manifest_ref, font_manifest_bytes = production_font_manifest(html_bytes.decode("utf-8"), asset_root)
    full_bytes = full_path.read_bytes()
    compact_bytes = compact_path.read_bytes()
    capture = json.loads(capture_path.read_text(encoding="utf-8"))
    capture_by_kind = {
        item.get("kind"): item
        for item in capture.get("derivatives", [])
        if isinstance(item, dict)
    }
    expected = {
        "full": (direction["preview_ref"], 2488, 1056, sha256_bytes(full_bytes)),
        "compact_622": (direction["compact_preview_ref"], 622, 264, sha256_bytes(compact_bytes)),
    }
    if capture.get("source_sha256") != sha256_bytes(html_bytes):
        raise RuntimeError("Capture report does not bind the selected HTML bytes.")
    for kind, (_, width, height, digest) in expected.items():
        item = capture_by_kind.get(kind)
        if not isinstance(item, dict) or (item.get("width"), item.get("height"), item.get("sha256")) != (width, height, digest):
            raise RuntimeError(f"Capture report does not bind the {kind} derivative.")
        painted_ratio = item.get("painted_ratio")
        if not isinstance(painted_ratio, (int, float)) or isinstance(painted_ratio, bool) or painted_ratio < 0.006:
            raise RuntimeError(f"Capture report marks the {kind} derivative as visually blank.")

    source_refs = unique([
        str(source_ref)
        for binding in bindings
        for source_ref in binding.get("source_refs", [])
    ])
    if not source_refs:
        raise RuntimeError("Selected direction bindings do not retain source refs.")

    message = direction_set["message"]
    grammar = direction_set["logic_progression"]["pattern"]
    color_system = direction["layout_system"]["color_system"]
    direction_id = direction["direction_id"]
    visual_suffix = direction_id.removeprefix("VDIR_")
    html_digest = sha256_bytes(html_bytes)
    tags = unique([str(message["direction"]), str(direction["route"])])
    refs_by_kind = {
        kind: [item["binding_id"] for item in bindings if item.get("kind") == kind]
        for kind in ("series", "metric", "level", "event")
    }

    return {
        "schema_version": "viewpoint-visual-v1",
        "visual_id": f"VVIS_{visual_suffix}_{html_digest[-12:]}",
        "render_profile": "wide_2488",
        "spec_ref": direction_id,
        "grammar": grammar,
        "payload_mode": payload_mode(bindings),
        "visual_job": "render_selected_direction",
        "state": state,
        "generated_at": generated_at,
        "dimensions": {"width": 2488, "height": 1056},
        "theme": color_system["preset_id"],
        "lineage": {
            "input_artifact_refs": unique([direction_set["direction_set_id"], *direction_set["input_refs"]]),
            "source_refs": source_refs,
            "series_refs": refs_by_kind["series"],
            "value_refs": refs_by_kind["metric"],
            "level_refs": refs_by_kind["level"],
            "event_refs": refs_by_kind["event"],
            "node_refs": [],
            "edge_refs": [],
            "rail_refs": [],
            "stage_refs": [],
            "decision_cutoff_at": decision_cutoff_at,
        },
        "content": {
            "headline": message["claim"],
            "observation": message["because"],
            "observed_at": observed_at,
            "strategy_tags": tags,
            "alt_text": f"{message['claim']}. {message['because']}. {message['implication']}.",
            "watermark": "Cuebook",
        },
        "asset": {
            "html": {"ref": direction["html_ref"], "sha256": html_digest},
            "svg": None,
            "font_manifest": {"ref": font_manifest_ref, "sha256": sha256_bytes(font_manifest_bytes)},
            "png_derivatives": [
                {"kind": kind, "ref": ref, "width": width, "height": height, "sha256": digest}
                for kind, (ref, width, height, digest) in expected.items()
            ],
            "derivative_bundle_hash": sha256_bytes(full_bytes + compact_bytes),
        },
        "quality_report": {"decision": "ready", "warnings": [], "hard_failures": []},
    }


def atomic_write(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary = path.with_name(f".{path.name}.{os.getpid()}.tmp")
    temporary.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    os.replace(temporary, path)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("direction_set", type=Path)
    parser.add_argument("--asset-root", type=Path, required=True)
    parser.add_argument("--observed-at", required=True)
    parser.add_argument("--decision-cutoff-at", required=True)
    parser.add_argument("--generated-at", required=True)
    parser.add_argument("--state", choices=("ready", "frozen"), default="frozen")
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    try:
        direction_set = json.loads(args.direction_set.read_text(encoding="utf-8"))
        direction_errors = load_direction_validator().validate(direction_set, args.asset_root)
        if direction_errors:
            details = "; ".join(f"{item['code']}: {item['message']}" for item in direction_errors)
            raise RuntimeError(f"Invalid VisualDirectionSetV1: {details}")
        manifest = build_manifest(
            direction_set,
            args.asset_root,
            observed_at=args.observed_at,
            decision_cutoff_at=args.decision_cutoff_at,
            generated_at=args.generated_at,
            state=args.state,
        )
        validation = validate_manifest(manifest, args.asset_root)
        if not validation["valid"]:
            details = "; ".join(f"{item['code']}: {item['message']}" for item in validation["errors"])
            raise RuntimeError(f"Final ViewpointVisualV1 failed validation: {details}")
        atomic_write(args.output, manifest)
    except (OSError, KeyError, json.JSONDecodeError, RuntimeError) as exc:
        print(json.dumps({"ok": False, "error": str(exc)}, ensure_ascii=False, indent=2))
        return 1
    print(json.dumps({"ok": True, "manifest": str(args.output), "visual_id": manifest["visual_id"]}, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/cuebook/public-skills/create-cuebook-content/references/modules/render-cuebook-viewpoint-visual/scripts/finalize_wide_viewpoint.mjs
#!/usr/bin/env node
// Freeze a selected VisualDirectionSetV1 direction as a wide ViewpointVisualV1.

import { createHash } from "node:crypto";
import { mkdirSync, readFileSync, realpathSync, renameSync, statSync, writeFileSync } from "node:fs";
import path, { basename, dirname, join, resolve } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

import { validate as validateDirectionSet } from "../../direct-cuebook-viewpoint-visual/scripts/validate_visual_direction_set.mjs";
import { validateManifest } from "./validate_viewpoint_visual.mjs";

const here = fileURLToPath(new URL(".", import.meta.url));
export const ROOT = resolve(here, "..");

function isFile(filePath) {
  try {
    return statSync(filePath).isFile();
  } catch {
    return false;
  }
}

export function sha256Bytes(data) {
  return `sha256:${createHash("sha256").update(data).digest("hex")}`;
}

function resolvedRoot(assetRoot) {
  try {
    return realpathSync(assetRoot);
  } catch {
    return resolve(assetRoot);
  }
}

export function safeAssetPath(assetRoot, ref, { base = null } = {}) {
  if (path.isAbsolute(ref) || ref.split("/").includes("..")) throw new Error(`Unsafe artifact-local asset ref: ${ref}`);
  const root = resolvedRoot(assetRoot);
  const lexical = resolve(base ?? assetRoot, ref);
  let candidate;
  try {
    candidate = realpathSync(lexical);
  } catch {
    candidate = lexical;
  }
  const relative = path.relative(root, candidate);
  if (relative === ".." || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) throw new Error(`Asset ref escapes the artifact root: ${ref}`);
  return candidate;
}

function escapeRegex(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function productionFontManifest(htmlText, assetRoot) {
  const attribute = (name) => {
    const match = new RegExp(`\\b${escapeRegex(name)}=["']([^"']+)["']`, "iu").exec(htmlText);
    if (!match) throw new Error(`Selected HTML is missing ${name}.`);
    return match[1];
  };
  if (attribute("data-font-profile") !== "cuebook-noi-v1") throw new Error("Selected HTML must use the cuebook-noi-v1 font profile.");
  if (attribute("data-font-license-mode") !== "production") throw new Error("Final publication requires production font license mode.");
  const manifestRef = attribute("data-font-manifest-ref");
  const manifestPath = safeAssetPath(assetRoot, manifestRef);
  if (!isFile(manifestPath)) throw new Error(`Font manifest is missing: ${manifestPath}`);
  const manifestBytes = readFileSync(manifestPath);
  let manifest;
  try {
    manifest = JSON.parse(manifestBytes.toString("utf8"));
  } catch {
    throw new Error(`Font manifest is invalid JSON: ${manifestPath}`);
  }
  if (manifest.schema_version !== "cuebook-font-assets-v1" || manifest.font_profile_id !== "cuebook-noi-v1") throw new Error("Font manifest does not bind the Cuebook Noi profile.");
  const licenseRef = String(manifest.license_ref || "");
  if (manifest.license_mode !== "production" || manifest.release_eligible !== true) throw new Error("Font manifest is not release-eligible production material.");
  if ([...licenseRef].length < 6 || /trial|eval/iu.test(licenseRef)) throw new Error("Production font manifest needs an opaque non-evaluation license_ref.");
  const cssRef = String(manifest.css_ref || ""), cssPath = safeAssetPath(assetRoot, cssRef, { base: dirname(manifestPath) });
  if (!isFile(cssPath) || sha256Bytes(readFileSync(cssPath)) !== manifest.css_sha256) throw new Error("Font CSS is missing or does not match the manifest hash.");
  const expectedCssRef = path.posix.join(path.posix.dirname(manifestRef), cssRef);
  if (!new RegExp(`href=["'](?:\\./)?${escapeRegex(expectedCssRef)}["']`, "iu").test(htmlText)) throw new Error("Selected HTML does not load the CSS bound by its font manifest.");
  const files = manifest.files;
  const weights = Array.isArray(files) ? new Set(files.filter((item) => item !== null && typeof item === "object" && !Array.isArray(item)).map((item) => item.weight)) : new Set();
  if (!Array.isArray(files) || weights.size !== 4 || ![400, 500, 600, 700].every((weight) => weights.has(weight))) throw new Error("Font manifest must bind upright Noi weights 400, 500, 600, and 700.");
  for (const item of files) {
    const ref = String(item.ref || ""), sourceName = String(item.source_name || "");
    if (/trial/iu.test(ref + sourceName)) throw new Error("Production font manifest cannot reference Trial font assets.");
    const fontPath = safeAssetPath(assetRoot, ref, { base: dirname(manifestPath) });
    if (!isFile(fontPath) || sha256Bytes(readFileSync(fontPath)) !== item.sha256) throw new Error(`Font asset is missing or does not match its manifest hash: ${ref}`);
  }
  return [manifestRef, manifestBytes];
}

export function unique(values) {
  return [...new Set(values.filter(Boolean))];
}

export function selectedDirection(directionSet) {
  if (directionSet.state !== "selected") throw new Error("VisualDirectionSetV1 must be selected before finalization.");
  const selectedId = directionSet.selected_direction_id;
  for (const direction of directionSet.directions ?? []) if (direction.direction_id === selectedId) return direction;
  throw new Error("Selected direction does not resolve inside VisualDirectionSetV1.");
}

export function payloadMode(bindings) {
  const kinds = new Set(bindings.map((item) => item.kind));
  const hasSeries = kinds.has("series");
  const hasKeyNumbers = ["metric", "level"].some((kind) => kinds.has(kind)) || bindings.some((item) => typeof item.value === "number");
  const hasQualitative = [...kinds].some((kind) => !new Set(["series", "metric", "level"]).has(kind));
  if (hasSeries && (hasKeyNumbers || hasQualitative)) return "mixed";
  if (hasSeries) return "series";
  if (hasKeyNumbers) return "key_numbers";
  return "qualitative";
}

export function buildManifest(directionSet, assetRoot, { observedAt, decisionCutoffAt, generatedAt, state = "frozen" }) {
  const direction = selectedDirection(directionSet);
  const bindingIds = new Set(direction.binding_refs ?? []);
  const bindings = (directionSet.bindings ?? []).filter((item) => bindingIds.has(item.binding_id));
  if (!bindings.length) throw new Error("Selected direction has no resolved bindings.");
  const htmlPath = join(assetRoot, direction.html_ref), fullPath = join(assetRoot, direction.preview_ref), capturePath = join(assetRoot, direction.capture_report_ref);
  for (const filePath of [htmlPath, fullPath, capturePath]) if (!isFile(filePath)) throw new Error(`Required selected-direction asset is missing: ${filePath}`);

  const htmlBytes = readFileSync(htmlPath);
  const [fontManifestRef, fontManifestBytes] = productionFontManifest(new TextDecoder("utf-8", { fatal: true }).decode(htmlBytes), assetRoot);
  const fullBytes = readFileSync(fullPath);
  const capture = JSON.parse(readFileSync(capturePath, "utf8"));
  const captureByKind = Object.fromEntries((capture.derivatives ?? []).filter((item) => item !== null && typeof item === "object" && !Array.isArray(item)).map((item) => [item.kind, item]));
  const expected = {
    full: [direction.preview_ref, 2488, 1056, sha256Bytes(fullBytes)],
  };
  if (capture.source_sha256 !== sha256Bytes(htmlBytes)) throw new Error("Capture report does not bind the selected HTML bytes.");
  for (const [kind, [, width, height, digest]] of Object.entries(expected)) {
    const item = captureByKind[kind];
    if (item === null || typeof item !== "object" || Array.isArray(item) || item.width !== width || item.height !== height || item.sha256 !== digest) throw new Error(`Capture report does not bind the ${kind} derivative.`);
    if (typeof item.painted_ratio !== "number" || item.painted_ratio < 0.006) throw new Error(`Capture report marks the ${kind} derivative as visually blank.`);
  }

  const sourceRefs = unique(bindings.flatMap((binding) => (binding.source_refs ?? []).map(String)));
  if (!sourceRefs.length) throw new Error("Selected direction bindings do not retain source refs.");
  const message = directionSet.message, grammar = directionSet.logic_progression.pattern, colorSystem = direction.layout_system.color_system, directionId = direction.direction_id;
  const visualSuffix = directionId.startsWith("VDIR_") ? directionId.slice("VDIR_".length) : directionId;
  const htmlDigest = sha256Bytes(htmlBytes), tags = unique([String(message.direction), String(direction.route)]);
  const refsByKind = Object.fromEntries(["series", "metric", "level", "event"].map((kind) => [kind, bindings.filter((item) => item.kind === kind).map((item) => item.binding_id)]));
  return {
    schema_version: "viewpoint-visual-v1",
    visual_id: `VVIS_${visualSuffix}_${htmlDigest.slice(-12)}`,
    render_profile: "wide_2488",
    spec_ref: directionId,
    grammar,
    payload_mode: payloadMode(bindings),
    visual_job: "render_selected_direction",
    state,
    generated_at: generatedAt,
    dimensions: { width: 2488, height: 1056 },
    theme: colorSystem.preset_id,
    lineage: {
      input_artifact_refs: unique([directionSet.direction_set_id, ...directionSet.input_refs]),
      source_refs: sourceRefs,
      series_refs: refsByKind.series,
      value_refs: refsByKind.metric,
      level_refs: refsByKind.level,
      event_refs: refsByKind.event,
      node_refs: [],
      edge_refs: [],
      rail_refs: [],
      stage_refs: [],
      decision_cutoff_at: decisionCutoffAt,
    },
    content: {
      headline: message.claim,
      observation: message.because,
      observed_at: observedAt,
      strategy_tags: tags,
      alt_text: `${message.claim}. ${message.because}. ${message.implication}.`,
      watermark: "Cuebook",
    },
    asset: {
      html: { ref: direction.html_ref, sha256: htmlDigest },
      svg: null,
      font_manifest: { ref: fontManifestRef, sha256: sha256Bytes(fontManifestBytes) },
      png_derivatives: Object.entries(expected).map(([kind, [ref, width, height, digest]]) => ({ kind, ref, width, height, sha256: digest })),
      derivative_bundle_hash: sha256Bytes(fullBytes),
    },
    quality_report: { decision: "ready", warnings: [], hard_failures: [] },
  };
}

export function atomicWrite(pathName, payload) {
  mkdirSync(dirname(pathName), { recursive: true });
  const temporary = join(dirname(pathName), `.${basename(pathName)}.${process.pid}.tmp`);
  writeFileSync(temporary, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  renameSync(temporary, pathName);
}

function parseCli(argv) {
  const result = { directionSet: null, assetRoot: null, observedAt: null, decisionCutoffAt: null, generatedAt: null, state: "frozen", output: null };
  const keys = { "--asset-root": "assetRoot", "--observed-at": "observedAt", "--decision-cutoff-at": "decisionCutoffAt", "--generated-at": "generatedAt", "--state": "state", "--output": "output" };
  for (let index = 0; index < argv.length; index += 1) {
    if (Object.hasOwn(keys, argv[index])) {
      result[keys[argv[index]]] = argv[index + 1] ?? null;
      index += 1;
    } else if (result.directionSet === null) result.directionSet = argv[index];
  }
  return result;
}

export function main(argv = process.argv.slice(2)) {
  const args = parseCli(argv);
  let output;
  try {
    if ([args.directionSet, args.assetRoot, args.observedAt, args.decisionCutoffAt, args.generatedAt, args.output].some((item) => item === null)) throw new Error("direction_set, asset timestamps, and output are required");
    if (!new Set(["ready", "frozen"]).has(args.state)) throw new Error("state must be ready or frozen");
    const directionSet = JSON.parse(readFileSync(args.directionSet, "utf8"));
    const directionErrors = validateDirectionSet(directionSet, args.assetRoot);
    if (directionErrors.length) throw new Error(`Invalid VisualDirectionSetV1: ${directionErrors.map((item) => `${item.code}: ${item.message}`).join("; ")}`);
    const manifest = buildManifest(directionSet, args.assetRoot, { observedAt: args.observedAt, decisionCutoffAt: args.decisionCutoffAt, generatedAt: args.generatedAt, state: args.state });
    const validation = validateManifest(manifest, args.assetRoot);
    if (!validation.valid) throw new Error(`Final ViewpointVisualV1 failed validation: ${validation.errors.map((item) => `${item.code}: ${item.message}`).join("; ")}`);
    atomicWrite(args.output, manifest);
    output = { ok: true, manifest: args.output, visual_id: manifest.visual_id };
  } catch (error) {
    output = { ok: false, error: error.message };
  }
  process.stdout.write(`${JSON.stringify(output, null, 2)}\n`);
  return output.ok ? 0 : 1;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) process.exitCode = main();
